### PR TITLE
Connect user to chat to increase watchtime

### PR DESCRIPTION
**Pull request by anoviel**
add irc chat client
list user as viewer in selected chat
viewtime increases while connected 

Referenced issue
https://github.com/gottagofaster236/Twitch-Channel-Points-Miner/issues/16

### CODE REVIEW

**Comment by Tkd-Alex** [comment]
I think there are a lot of things that should be changed. 
I'll start a review asap!

File: TwitchChannelPointsMiner/TwitchChannelPointsMiner.py [hunk 1]
```diff
@@ -260,6 +260,10 @@ def end(self, signum, frame):
         if self.twitch_browser is not None:
             self.twitch_browser.browser.quit()
 
+        for streamer in self.streamers:
+            if streamer is not None:
```
Comment (Tkd-Alex) [inline TwitchChannelPointsMiner/TwitchChannelPointsMiner.py:264]: We don't need `if streamer is not None:`
Why a streamer should be None?

File: TwitchChannelPointsMiner/classes/entities/Chat.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+
+class Chat(irc.bot.SingleServerIRCBot):
+    def __init__(self, username, token, channel):
+        self.token = token
+        self.channel = "#" + channel
+
+        # Create IRC bot connection
+        server = "irc.chat.twitch.tv"
+        port = 6667
+        irc.bot.SingleServerIRCBot.__init__(
```
Comment (Tkd-Alex) [inline TwitchChannelPointsMiner/classes/entities/Chat.py:12]: - `server = "irc.chat.twitch.tv"`
- `port = 6667`

Should be moved in `constants/twitch.py`

File: TwitchChannelPointsMiner/classes/TwitchChat.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class ChatSettings(object):
+    def __init__(
+        self,
+        username: str = None,
+        token: str = None,
+    ):
+        self.username = username
+        self.token = token
+
```
Comment (Tkd-Alex) [inline TwitchChannelPointsMiner/classes/TwitchChat.py:17]: We don't need the class `ChatSettings`
It's only `self.username` and `self.token`? 

We have already this value.

File: TwitchChannelPointsMiner/classes/entities/Streamer.py [hunk 4]
```diff
@@ -18,12 +19,14 @@ def __init__(
         claim_drops: bool = None,
         watch_streak: bool = None,
         bet: BetSettings = None,
+        chat_client: ChatSettings = None,
```
Comment (Tkd-Alex) [inline TwitchChannelPointsMiner/classes/entities/Streamer.py:22]: as already said we don't need: `ChatSettings`

**Review by Tkd-Alex — CHANGES_REQUESTED** [review]

**Comment by Tkd-Alex** [comment]
If you want you can take a look of this https://pypi.org/project/python-twitch-irc/

File: TwitchChannelPointsMiner/classes/TwitchChat.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+class TwitchChat(Process):
```
Comment (Tkd-Alex) [inline TwitchChannelPointsMiner/classes/TwitchChat.py:19]: It's a long running-job with lower I/O.
We should extend a `Thread` instead of `Process`

**Review by Tkd-Alex — CHANGES_REQUESTED** [review]

**Comment by anoviel** [comment]
I had some issues getting pre-commit up and running on windows 😅
Thanks for the feedback, I will improve things as soon as possible :)

**Comment by Tkd-Alex** [comment]
> I had some issues getting pre-commit up and running on windows 
> Thanks for the feedback, I will improve things as soon as possible :)

If you have a problem with pre-commit run the following commands:
- `pre-commit uninstall`
- `pip install pyflakes, black, isort`
- `isort TwitchChannelPointsMiner/`
- `black TwitchChannelPointsMiner/`
- `flake8 TwitchChannelPointsMiner/`\

So you can check the code manually :smiley:

**Comment by Tkd-Alex** [comment]
@anoviel Have you the time for work on it? or do you prefer that I complete the review?

**Review by Tkd-Alex — APPROVED** [review]

**Comment by Tkd-Alex** [comment]
Hi :) After a long time I've updated this PR.
If someone wants to try :smile_cat:

**Comment by kelukelol** [comment]
> Hi :) After a long time I've updated this PR.
> If someone wants to try 😸

Going to test it for an hour or so. I will let you know if something comes up :)
image

**Comment by kelukelol** [comment]
`10/03/21 03:42:17 - INFO - [__print_report]: Duration 1:21:55.017667` works just fine without any issues appearing.

**Comment by Tkd-Alex** [comment]
The script workings as well with on-line/offline events? Because currently I don't know what happened when a streamer go offline and then online again

**Comment by kelukelol** [comment]
I am not sure about that too. I have not seen this happen in 2 hours just looking at the script work :/

**Comment by PureFallen** [comment]
Someone could also just open a Private Window/Incognito Tab and check in the Twitch Chat´s Memberlist, if the Accountname is listed. If so, the feature works as intended. It even works for more than 2 streamers, given that the viewlimit does not apply to Twitch Chat. Glad that this was considered in the PR.

However, the new feature draws attention to 2 things:

- Dealing with channels that don't have Channel Points active
It might be attractive for someone that migrates from another solutions to this software to join the IRC from channels that don't support Channel Points, regardless of the intentional purpose of the software. The reasons for that can vary, like the Watchtime example @anoviel mentioned before, farming alternative currencies such as StreamElements/StreamHeroes (which are measured by time active in chat), qualifying for drops or subscription gifts (which select random people that were in the chatbox in the last x hours) or more.

People could put those streams simply on the very bottom of the list, but assuming every other stream in the list is offline, they might up creating load on script and network for farming Viewtime on a stream it is not worth it... unless there is a way to detect that a channel doesn't support Channel Points?
An alternative approach would be to provide a setting `irc_only` (default: false) for the Streamer Settings, automatically making them treat with lowest priority and completely ignored for the requests related for Channel Points, Predictions.... stuff like that.

- Dealing with channels that reward people for being in the chatbox, while the Stream if offline
This sounds really odd in the first moment, but there are Streamer which are actually checking and rewarding people for the time they spend in Chatbox regardless of the Stream's Online State. The reasons for that can vary over wanting to maintain an active chat, even if offline (to make the impression of a healthy/active community) or sometimes even easier: using a selfcoded bot and messing up the Online-Check for it.

Once again there would probably be a setting required to join Streamer´s IRC regardless of their Online State, but at this point it might be even better for the readability and usability to have a seperate list of Streamers in the run.py where people can put streams that are exclusive for IRC usage and configure the behaviour with Offline Streams there, before the "StreamerSettings" in `twitch_miner.mine` grow even longer...

**Comment by Yishaqel** [comment]
The latest log (which has been running since last night) takes 100MB because it is saving the messages in the IRC channels. I don't think I can share it but I have received a couple of errors today.

**Comment by Tkd-Alex** [comment]
> The latest log (which has been running since last night) takes 100MB because it is saving the messages in the IRC channels. I don't think I can share it but I have received a couple of errors today.

They are all messages from irc chat. 
I'll remove asap because we don't need them.

**Comment by kelukelol** [comment]
Keep getting an issue every time the bot leaves the IRC chat.
image 
I bet this has something to do with me having Windows :/

**Comment by Tkd-Alex** [comment]
:cry: :cry: :cry: 

On Linux for the moment no problem when the streamer goes offline.
image

**Comment by Tkd-Alex** [comment]
Are you able to edit add a time sleep here?
https://github.com/anoviel/Twitch-Channel-Points-Miner-v2/blob/46918fda9b181af2fa5243587cdf9256b4fcc129/TwitchChannelPointsMiner/classes/entities/Chat.py#L26-L27

```python
import time
....
code
code
....
self.reactor.process_once(timeout=0.2)
time.sleep(0.01)
```

We could also surround with `try except` and see if something change.

**Comment by kelukelol** [comment]
Bruh, it just fixed itself without me making any changes.

**Comment by Tkd-Alex** [comment]
Awesome! :rofl: 
Anyway I think I'll do this changes just for prevent future errors.

**Comment by PureFallen** [comment]
> 
> > unless there is a way to detect that a channel doesn't support Channel Points?
> 
> I think yes, should be a way. If you can provide a streamer without channel points maybe I can't find a way.
> 

Not want to advertise Streamers, but you could take a look into Twitch's "Slot" Category. Streamer which exclusively stream Slots don't meet the requirements for Affiliation/Partnership and therefore can't enable Channel Points.
You will find someone really quickly there^^

> uhm agree, your message make sense. I got your points.
> Currently, I think I'll merge this PR asap and then create a new branch with your idea and proposals. Because seems that @anoviel haven't free time to work on It and for me, It's a little uncomfortable to work on another user's repository. We could consider [#47 (comment)](https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/47#issuecomment-795499841) as improvements.

Oh, sure. I am just extremly unexperienced with Github in general. Should learn that at some point. Limiting it to theorycrafting most of the time for now. I really appreciate that.

**Comment by Tkd-Alex** [comment]
image

Here another error :cry: 
I'm currently working on this.

**Comment by Tkd-Alex** [comment]
Seems to be stable.
I'll merge asap, anyone has founded problems? 

@PureFallen after the merge please open a new issue. You can easily do a Copy & Paste or open new issue directly via GitHub.

image

**Comment by kelukelol** [comment]
> Seems to be stable.
> I'll merge asap, anyone has founded problems?
> 
> @PureFallen after the merge please open a new issue. You can easily do a Copy & Paste or open new issue directly via GitHub.
> 
> image

Left the script running over this night. After reading logs, no issues have been found. (Despite this issue https://github.com/Tkd-Alex/Twitch-Channel-Points-Miner-v2/pull/47#issuecomment-795581741 but it fixed itself. )

**Pull request merged**